src/Week4/HigherOrder.hs

```diff
@@ -0,0 +1,181 @@
+            deriving (Show, Eq)
+
+height :: Tree a -> Integer
+height Leaf           = -1
+height (Node h _ _ _) = h
+
+
+insertNode :: a -> Tree a -> Tree a
+insertNode x Leaf                   = Node 0 Leaf x Leaf
+insertNode x (Node h left root right)
+  | h1 < h2   = Node h (insertNode x left) root right
+  | h1 > h2   = Node h left root (insertNode x right)
+  | otherwise = Node (h3 + 1) leftInsert root right
+  where h1 = height left
+        h2 = height right
+        h3 = height leftInsert
+        leftInsert = (insertNode x left)
+
+
+foldTree :: [a] -> Tree a
+foldTree = foldr insertNode Leaf
+
+
+{-
+Exercise 3: Implement a function
+   xor :: [Bool] -> Bool
+
+ which returns True if and only if there are an odd number of True
+ values contained in the input list. It does not matter how many
+ False values the input list contains.
+
+ For example,
+   xor [False, True, False] == True
+   xor [False, True, False, False, True] == False
+
+ Your solution must be implemented using a fold.
+-}
+-- XOR: result in each position is 1 if only one of the bits is 1,
+-- but will be 0 if both are 0 or both are 1.
+xor :: [Bool] -> Bool
+xor = foldr (/=) False
+
+-- The above is also possible using pattern matching
+xor' :: [Bool] -> Bool
+xor' = foldr xorUtil False
+  where xorUtil n m
+          | n == True && m == False = True
+          | n == False && m == True = True
+          | otherwise               = False
+
+
+{-
+ Exercise 4: Implement map as a fold. That is, complete the definition
+    map’ :: (a -> b) -> [a] -> [b]
+    map’ f = foldr ...
+
+ in such a way that map’ behaves identically to the standard map function.
+-}
+map' :: (a -> b) -> [a] -> [b]
+map' f = foldr (\x gs -> (f x) : gs) []
+
+
+{-
+ Exercise 4: Finding primes
+
+ Read about the Sieve of Sundaram. Implement the algorithm using function composition.
+
+ Given an integer n, your function should
+ generate all the odd prime numbers up to 2n + 2.
+   sieveSundaram :: Integer -> [Integer]
+   sieveSundaram = ...
+
+ To give you some help, below is a function to compute the Cartesian product
+ of two lists. This is similar to zip, but it produces all
+ possible pairs instead of matching up the list elements.
+
+ For example,
+   cartProd [1,2] [’a’,’b’] == [(1,’a’),(1,’b’),(2,’a’),(2,’b’)]
+
+ It’s written using a list comprehension, which we haven’t talked about
+ in class (but feel free to research them).
+   cartProd :: [a] -> [b] -> [(a, b)]
+   cartProd xs ys = [(x,y) | x <- xs, y <- ys]
+-}
+
+{-
+ Start with a list of the integers from 1 to n.
+ From this list, remove all numbers of the form i + j + 2ij where:
+    i, j belongs to N, 1 <= i <= j
+    i + j + 2ij <= n
+
+ The remaining numbers are doubled and incremented by one, giving a list
+ of the odd prime numbers (i.e., all primes except 2) below 2n + 1
+-}
+seq' :: Integer -> [Integer]
+seq' n = filter (<= n) [i + j + 2 * i * j | j <- [1..n], i <- [1..j], i + j + 2 * i * j <= n]
```

Comment:
It appears we are filtering twice here. The list comprehension has a predicate that filters the following: ` i + j + 2 * i * j <= n`

```suggestion
seq' n = [i + j + 2 * i * j | j <- [1..n], i <- [1..j], i + j + 2 * i * j <= n]
```

Reply:
Ah you're right! `filter` is just useless here since I do everything in the comprehension itself.